File: src/brs/grpc/handlers/SubmitNonceHandler.java
package brs.grpc.handlers;

import brs.Account;
import brs.Blockchain;
import brs.Generator;
import brs.crypto.Crypto;
import brs.grpc.GrpcApiHandler;
import brs.grpc.proto.ApiException;
import brs.grpc.proto.Brs;
import brs.services.AccountService;
import brs.util.Convert;

import java.util.Objects;

public class SubmitNonceHandler implements GrpcApiHandler<Brs.SubmitNonceRequest, Brs.SubmitNonceResponse> {

    private final Blockchain blockchain;
    private final AccountService accountService;
    private final Generator generator;

    public SubmitNonceHandler(Blockchain blockchain, AccountService accountService, Generator generator) {
        this.blockchain = blockchain;
        this.accountService = accountService;
        this.generator = generator;
    }

    @Override
    public Brs.SubmitNonceResponse handleRequest(Brs.SubmitNonceRequest request) throws Exception {
        String secret = request.getSecretPhrase();
        long nonce = request.getNonce();

        String accountId = Convert.toUnsignedLong(request.getAccount());

        int submissionHeight = request.getBlockHeight();

        if (submissionHeight != 0) {
            if (submissionHeight != blockchain.getHeight() + 1) {
                throw new ApiException("Given block height does not match current blockchain height");
            }
        }

        if (Objects.equals(secret, "")) {
            throw new ApiException("Missing Passphrase");
        }

        byte[] secretPublicKey = Crypto.getPublicKey(secret);
        Account secretAccount = accountService.getAccount(secretPublicKey);
        if(secretAccount != null) {
            Account genAccount;
            if(accountId != null) {
                genAccount = accountService.getAccount(Convert.parseAccountId(accountId));
            }
            else {
                genAccount = secretAccount;
            }

            if(genAccount != null) {
                Account.RewardRecipientAssignment assignment = accountService.getRewardRecipientAssignment(genAccount);
                long rewardId;
                if (assignment == null) {
                    rewardId = genAccount.getId();
                }
                else if (assignment.getFromHeight() > blockchain.getLastBlock().getHeight() + 1) {
                    rewardId = assignment.getPrevRecipientId();
                }
                else {
                    rewardId = assignment.getRecipientId();
                }
                if (rewardId != secretAccount.getId()) {
                    throw new ApiException("Passphrase does not match reward recipient");
                }
            }
            else {
                throw new ApiException("Passphrase is for a different account");
            }
        }

        Generator.GeneratorState generatorState;
        if (accountId == null || secretAccount == null) {
            generatorState = generator.addNonce(secret, nonce);
        }
        else {
            Account genAccount = accountService.getAccount(Convert.parseUnsignedLong(accountId));
            if (genAccount == null || genAccount.getPublicKey() == null) {
                throw new ApiException("Passthrough mining requires public key in blockchain");
            }
            else {
                byte[] publicKey = genAccount.getPublicKey();
                generatorState = generator.addNonce(secret, nonce, publicKey);
            }
        }

        if (generatorState == null) {
            throw new ApiException("Failed to create generator");
        }

        return Brs.SubmitNonceResponse.newBuilder().setDeadline(generatorState.getDeadline().longValueExact()).build();
    }
}

File: src/main/kotlin/brs/api/grpc/handlers/SubmitNonceHandler.kt
package brs.api.grpc.handlers

import brs.entity.Account
import brs.services.BlockchainService
import brs.services.GeneratorService
import brs.util.crypto.Crypto
import brs.api.grpc.GrpcApiHandler
import brs.api.grpc.service.ApiException
import brs.api.grpc.proto.BrsApi
import brs.services.PropertyService
import brs.objects.Props
import brs.services.AccountService
import burst.kit.crypto.BurstCrypto

class SubmitNonceHandler(propertyService: PropertyService, private val blockchainService: BlockchainService, private val accountService: AccountService, private val generatorService: GeneratorService) : GrpcApiHandler<BrsApi.SubmitNonceRequest, BrsApi.SubmitNonceResponse> {
    private val passphrases: Map<Long, String>
    private val allowOtherSoloMiners: Boolean

    init {

        this.passphrases = propertyService.get(Props.SOLO_MINING_PASSPHRASES).associateBy { passphrase -> BurstCrypto.getInstance().getBurstAddressFromPassphrase(passphrase).burstID.signedLongId }
        this.allowOtherSoloMiners = propertyService.get(Props.ALLOW_OTHER_SOLO_MINERS)
    }

    override fun handleRequest(request: BrsApi.SubmitNonceRequest): BrsApi.SubmitNonceResponse {
        var secret: String = request.secretPhrase
        val nonce = request.nonce
        val accountId = request.account
        val submissionHeight = request.blockHeight

        if (submissionHeight != 0 && submissionHeight != blockchainService.height + 1) {
            throw ApiException("Given block height does not match current blockchain height")
        }

        if (secret.isEmpty()) {
            if (passphrases.containsKey(accountId)) {
                secret = passphrases[accountId]!!
            } else {
                throw ApiException("Missing Passphrase and account passphrase not in solo mining config")
            }
        }

        if (!allowOtherSoloMiners && !passphrases.containsValue(secret)) {
            throw ApiException("This account is not allowed to mine on this node as the whitelist is enabled and it is not whitelisted.")
        }

        val secretPublicKey = Crypto.getPublicKey(secret)
        val secretAccount = accountService.getAccount(secretPublicKey)
        if (secretAccount != null) {
            verifySecretAccount(accountService, blockchainService, secretAccount, accountId)
        }

        val generatorState: GeneratorService.GeneratorState?
        generatorState = if (accountId == 0L || secretAccount == null) {
            generatorService.addNonce(secret, nonce)
        } else {
            val genAccount = accountService.getAccount(accountId)
            if (genAccount?.publicKey == null) {
                throw ApiException("Passthrough mining requires public key in blockchain")
            } else {
                val publicKey = genAccount.publicKey
                generatorService.addNonce(secret, nonce, publicKey!!)
            }
        }

        if (generatorState == null) {
            throw ApiException("Failed to create generator")
        }

        return BrsApi.SubmitNonceResponse.newBuilder().setDeadline(generatorState.deadline.longValueExact()).build()
    }

    companion object {

        fun verifySecretAccount(accountService: AccountService, blockchainService: BlockchainService, secretAccount: Account, accountId: Long) {
            val genAccount = if (accountId != 0L) {
                accountService.getAccount(accountId)
            } else {
                secretAccount
            }

            if (genAccount != null) {
                val assignment = accountService.getRewardRecipientAssignment(genAccount)
                val rewardId = when {
                    assignment == null -> genAccount.id
                    assignment.fromHeight > blockchainService.lastBlock.height + 1 -> assignment.prevRecipientId
                    else -> assignment.recipientId
                }
                if (rewardId != secretAccount.id) {
                    throw ApiException("Passphrase does not match reward recipient")
                }
            } else {
                throw ApiException("Passphrase is for a different account")
            }
        }
    }
}